### DISCUSSION TITLE
Typescript and large Root store - The inferred type of this node exceeds the maximum length the compiler will serialize.

### DISCUSSION THREAD

**Discussion by sergioisidoro** [mobxjs/mobx-state-tree#2115] [Q&A] [answered]
I'm hitting a problem as my stores and models keep getting bigger.

```
The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.ts(7056)
```

I have checked and I can't find any recursive declaration that could result in this. I'm actually hitting the limit just by adding some additional simple field declarations (the error goes away if I comment it)

Wondering if anyone has came across this problem, and how did you solve it? And also, how did you debug it (how did you see if there's any leaky declaration of the type?)

**Comment by coolsoftwaretyler** [comment]
Hey @sergioisidoro - I've never seen this before. Sounds really frustrating, but also kind of neat (in an annoying way).

What I do know is that our existing TypeScript types are somewhat verbose, so this doesn't shock me. Is there any chance you can share:

1. Your TS version
2. The definition of your root store model?

**Reply by sergioisidoro** [reply]
Hey! Thanks for replying. :)

After a while more chacing my tail, I finally found the reason. I used ignite boilerplate to start my project. It has this nice helper to avoid doing setter actions which can be a bit annoyed.
```
export const withSetPropAction = <T extends IStateTreeNode>(mstInstance: T) => ({
  // generic setter for all properties
  setProp<K extends keyof SnapshotIn<T>, V extends SnapshotIn<T>[K]>(field: K, newValue: V) {
    // @ts-ignore - for some reason TS complains about this, but it still works fine
    mstInstance[field] = newValue
  },
})
```

This works great at first, but if your tree in deep or is deeply nested, the type definitions for this method grow dramatically. I just happened to be exactly on the limit that adding a few single new attributes would break the camel's back.

I think this might be related to the children nodes also having a setProp method, because that was the only method signature being included in the type recursively. None of the other actions did.

To debug this:
1. Remove parts of the root store untill typescript stops complaining
2. Build your project
3. Inspect your types.d.ts file to check for nestes types

I also files an issue with the boilerplate project 
https://github.com/infinitered/ignite/discussions/2547

**Reply by coolsoftwaretyler** [reply]
Makes sense, @sergioisidoro!

I think the root cause here is that MobX-State-Tree TypeScript types are somewhat verbose, and the utilities offered by Ignite are built on top of a type system that can get out of hand pretty easily.

I don't have a great path forward to fix the Types. We are actively discussing and thinking about them, but I have no ETA on when they'll be better, and I don't think Ignite will be able to help you much on their end, either. You might consider:

1. Implementing their utility in your own codebase, but with some simpler typing information, and see if that helps
2. It's possible you could find a way to un-nest your nested data model. I'd be more than happy to chat about it here if you're able to post (I totally understand if you're unable or unwilling to do that).

I'm sorry that my best advice here boils down to "fix it yourself and good luck!" - I wish I could be more helpful. Again, I'm happy to talk through the details, but I do want to be honest and realistic that you've probably hit a limitation of our underlying TypeScript implementation. I don't know if MST or Ignite will have a straightforward path to resolve this for you.

**Reply by sergioisidoro** [reply, marked as answer]
No problem :) 

My expectations about opening this thread was to allow others to find this problem if they were struggling.
In my case I managed to delay the problem by removing the `withSetProp` action from the models that don't use it, which pruned the type declarations enough for the typescript compiler to stop complaining. 

If this ever becomes a problem again, I'll probably have to tweak those utilities and maybe pass the type annotations some other way to those generic actions (for example, by explicitly passing they keys of which setters I want to generate, and what are their types).

Thanks for taking the time to answer this. If I find some better solution in the future I'll post it here as a reference to others.

**Reply by coolsoftwaretyler** [reply]
Sounds good! I'm going to mark your answer as "the answer" here, but yes please write up any additional findings you have. Really appreciate it!
